x/evm/keeper/msg_assigner.go

```diff
@@ -3,266 +3,324 @@ package keeper
-	highScore := 0
-	var highScorers []string
-
-	// Get the validators with the highest score.
-	for addr, score := range scored {
-		switch {
-		case score > highScore:
-			highScore = score
-			highScorers = []string{addr}
-		case score == highScore:
-			highScorers = append(highScorers, addr)
-		}
-	}
+	ranked := make([]validatorScore, 0, len(validatorsInfos))
 
-	// All else equal, grab one of our high scorers, but not always the same one
-	sort.Strings(highScorers)
-	return highScorers[int(sdkCtx.BlockHeight())%len(highScorers)]
-}
+	numValidators := len(validatorsInfos)
 
-func (ma MsgAssigner) PickValidatorForMessage(ctx context.Context, weights *types.RelayWeights, chainID string, req *xchain.JobRequirements) (string, error) {
-	currentSnapshot, err := ma.ValsetKeeper.GetCurrentSnapshot(ctx)
-	if err != nil {
-		return "", err
+	infosSlice := make([]ValidatorInfo, 0, numValidators)
+	for _, info := range validatorsInfos {
+		infosSlice = append(infosSlice, info)
 	}
 
-	// For some reason, GetCurrentSnapshot returns nil, nil when no shapshot exists.  We have to catch that and turn it into an error
-	if currentSnapshot == nil {
-		return "", errors.New("no snapshot found")
+	pdw := slice.Reduce[perfDataWindow, ValidatorInfo](infosSlice, func(window perfDataWindow, val ValidatorInfo) perfDataWindow {
+		window.minFee = probeMin(val.Fee, window.minFee)
+		window.maxFee = probeMax(val.Fee, window.maxFee)
+		window.minUptime = probeMin(val.Uptime, window.minUptime)
+		window.maxUptime = probeMax(val.Uptime, window.maxUptime)
+		window.minSuccessRate = probeMin(val.SuccessRate, window.minSuccessRate)
+		window.maxSuccessRate = probeMax(val.SuccessRate, window.maxSuccessRate)
+		window.minExecutionTime = probeMin(val.ExecutionTime, window.minExecutionTime)
+		window.maxExecutionTime = probeMax(val.ExecutionTime, window.maxExecutionTime)
+		window.minFeatureSet = probeMin(val.FeatureSet, window.minFeatureSet)
+		window.maxFeatureSet = probeMax(val.FeatureSet, window.maxFeatureSet)
+		return window
+	})
+
+	if !pdw.assert() {
+		// If we can't assert the metadata, we can't rank the validators.
+		return scoreSnapshot{}, errors.New("performance data window assertion failed")
 	}
 
-	assignableValidators := filterAssignableValidators(currentSnapshot.Validators, chainID, req)
-	if len(assignableValidators) == 0 {
-		return "", errors.New("no assignable validators for message")
+	for valAddr, info := range validatorsInfos {
+		scores := map[string]math.LegacyDec{
+			"fee":           scoreValue(*pdw.maxFee, *pdw.minFee, info.Fee, true),
+			"uptime":        scoreValue(*pdw.maxUptime, *pdw.minUptime, info.Uptime, false),
+			"successRate":   scoreValue(*pdw.maxSuccessRate, *pdw.minSuccessRate, info.SuccessRate, false),
+			"executionTime": scoreValue(*pdw.maxExecutionTime, *pdw.minExecutionTime, info.ExecutionTime, true),
+			"featureSet":    scoreValue(*pdw.maxFeatureSet, *pdw.minFeatureSet, info.FeatureSet, false),
+		}
```

Comment:
Might just be personal preference, but why use a map instead of individual variables?

Reply:
This was from the original implementation, I can only assume the engineer back then was looking for something like dynamic flexibility with these metrics.

I just didn't touch it, but I agree this could be moved to a struct.